src/components/ApplicationContext/ApplicationContext.tsx

```diff
@@ -133,6 +150,48 @@ export const ApplicationProvider: React.FC = () => {
     }
   }
 
+  const saveData = () => {
+    const { fname, lname, phone, email } = contactFormData
+    const {
+      age,
+      pronouns,
+      race,
+      sexuality,
+      school,
+      collegeAffiliation,
+      eventLocation,
+      major,
+      currentStanding,
+      country,
+    } = demographicFormData
+    const { whyCruzHacks, newThisYear, grandestInvention } = shortAnswerFormData
+    const { firstCruzHacks, hackathonCount, priorExperience } =
+      priorExperienceFormData
+    const { linkedin, github, cruzCoins, anythingElse } = connectedFormData
+    const { conductAgree, tosAgree } = mlhFormData
+
+    const data: SavedApplication = {
+      contact: { fname, lname, phone, email },
+      demographic: {
+        age,
+        pronouns,
+        race,
+        sexuality,
+        school,
+        collegeAffiliation,
+        eventLocation,
+        major,
+        currentStanding,
+        country,
+      },
+      shortAnswer: { whyCruzHacks, newThisYear, grandestInvention },
+      priorExperience: { firstCruzHacks, hackathonCount, priorExperience },
+      connected: { linkedin, github, cruzCoins, anythingElse },
+      MLH: { conductAgree, tosAgree },
+    }
+    store("application", data, 60 * 60 * 24 * 30, user?.email)
```

Comment:
I think one thing we should do is run the validation functions before saving. For example if a user inputs binary data, that will be stored in localstorage. If you run the validation functions with the onSubmit argument set to False it should validate only fields that must always be true (i.e valid alphanumeric data)

Reply:
I was able to save+load emojis/non standard chars

Reply:
That makes sense! Will do that after today's meeting.

Should we add a flash message above the form to notify the user if we can't save? It'll probably be useful for our submit function too, if the validation fails for the previous pages. Something like this for example:

<img src="https://user-images.githubusercontent.com/20177171/140810264-4d01ae93-a9d5-4464-af43-b0a9102c567e.png" width="500px" />(example from the Internet)

Reply:
Something like that could work! Could we also hold of some changes on this for now. I've been thinking about future proofing security for 2023 onwards and I wanted to get your thoughts on switching to session storage. If we do switch to session which we'll talk about at today's meeting, we can probably just run the save function onClick of the next/prev button.

Main reason I was thinking of switching is a scenario like this.

Localstorage: User Saves App-> Never submits app -> Data is never cleared -> user never revisits cruzhacks.com for 1 year -> auth0 is off -> stored personal data can never be cleared (unless user manually clears browsers cache) -> Potential XSS vulnerability in 2023 onward or even in current code base (data is compromised) or this could be a problem to future teams using caching (i.e similar storage conventions)

Reply:
This is definitely a valid concern. I also don't like storing the application in local storage.

Although, if we don't want to use local storage, perhaps we shouldn't store it at all? If we switch to session storage, I assume we would remove the "save" option so we don't confuse our user. But how often would a user navigate away from the "/portal" route (our ApplicationContext) and come back within the session?

I guess another option is to just ask for confirmation before the user navigates away from the portal view, similar to how Google Forms does it, and not bother with saving anything that is not permanent.

We should talk more about this during today's meeting.